llvm/test/CodeGen/AIE/aie2/GlobalISel/propagate-mmo-noptr.mir

```diff
@@ -0,0 +1,121 @@
+    ; CHECK-NEXT: [[MOVXM_lng_cg:%[0-9]+]]:ep_as_32bit = MOVXM_lng_cg target-flags(aie2-global) @softmax_ilut_ab
+    ; CHECK-NEXT: [[COPY2:%[0-9]+]]:er = COPY [[MOVXM_lng_cg]]
+    ; CHECK-NEXT: [[MOVXM_lng_cg1:%[0-9]+]]:ep_as_32bit = MOVXM_lng_cg target-flags(aie2-global) @softmax_ilut_cd
+    ; CHECK-NEXT: [[COPY3:%[0-9]+]]:er = COPY [[MOVXM_lng_cg1]]
+    ; CHECK-NEXT: [[MOVXM_lng_cg2:%[0-9]+]]:ep_as_32bit = MOVXM_lng_cg target-flags(aie2-global) @softmax_flut_ab
+    ; CHECK-NEXT: [[COPY4:%[0-9]+]]:er = COPY [[MOVXM_lng_cg2]]
+    ; CHECK-NEXT: [[MOVXM_lng_cg3:%[0-9]+]]:ep_as_32bit = MOVXM_lng_cg target-flags(aie2-global) @softmax_flut_cd
+    ; CHECK-NEXT: [[COPY5:%[0-9]+]]:er = COPY [[MOVXM_lng_cg3]]
+    ; CHECK-NEXT: [[COPY6:%[0-9]+]]:mdm = COPY [[COPY2]]
+    ; CHECK-NEXT: [[COPY7:%[0-9]+]]:er = COPY [[COPY6]]
+    ; CHECK-NEXT: [[COPY8:%[0-9]+]]:mdm = COPY [[COPY3]]
+    ; CHECK-NEXT: [[COPY9:%[0-9]+]]:er = COPY [[COPY8]]
+    ; CHECK-NEXT: [[MOV_RLC_imm10_pseudo:%[0-9]+]]:ers4 = MOV_RLC_imm10_pseudo 0
+    ; CHECK-NEXT: [[MOV_RLC_imm10_pseudo1:%[0-9]+]]:er = MOV_RLC_imm10_pseudo 2
+    ; CHECK-NEXT: [[MOV_RLC_imm10_pseudo2:%[0-9]+]]:er = MOV_RLC_imm10_pseudo 6
+    ; CHECK-NEXT: [[COPY10:%[0-9]+]]:mss = COPY [[COPY]]
+    ; CHECK-NEXT: [[VFLOOR_S32_BF16_mFl2FxSrc_W:%[0-9]+]]:vec512 = VFLOOR_S32_BF16_mFl2FxSrc_W [[COPY1]], [[COPY10]], implicit-def dead $srf2iflags, implicit $crf2imask
+    ; CHECK-NEXT: [[VSHUFFLE:%[0-9]+]]:vec512 = VSHUFFLE [[VFLOOR_S32_BF16_mFl2FxSrc_W]], [[VFLOOR_S32_BF16_mFl2FxSrc_W]], [[MOV_RLC_imm10_pseudo1]]
+    ; CHECK-NEXT: [[COPY11:%[0-9]+]]:ewl = COPY [[VSHUFFLE]].sub_256_lo
+    ; CHECK-NEXT: [[COPY12:%[0-9]+]]:vec256 = COPY [[COPY11]]
+    ; CHECK-NEXT: [[COPY13:%[0-9]+]]:mss = COPY [[MOV_RLC_imm10_pseudo]]
+    ; CHECK-NEXT: [[VUPS_S64_D16_mv_ups_w2c:%[0-9]+]]:acc1024 = VUPS_S64_D16_mv_ups_w2c [[COPY12]], [[COPY13]], implicit-def dead $srups_of, implicit $crsat, implicit $crupssign
+    ; CHECK-NEXT: [[COPY14:%[0-9]+]]:mss = COPY [[MOV_RLC_imm10_pseudo2]]
+    ; CHECK-NEXT: [[VSRS_S32_S64_mv_x_srs:%[0-9]+]]:vec512 = VSRS_S32_S64_mv_x_srs [[VUPS_S64_D16_mv_ups_w2c]], [[COPY14]], implicit-def dead $srsrs_of, implicit $crsat, implicit $crrnd
+    ; CHECK-NEXT: [[VBCST_32_:%[0-9]+]]:vec512 = VBCST_32 [[COPY7]]
+    ; CHECK-NEXT: [[VBCST_32_1:%[0-9]+]]:vec512 = VBCST_32 [[COPY9]]
+    ; CHECK-NEXT: [[MOVXM_lng_cg4:%[0-9]+]]:ers8 = MOVXM_lng_cg 52428
+    ; CHECK-NEXT: [[VSEL_32_:%[0-9]+]]:vec512 = VSEL_32 [[VBCST_32_]], [[VBCST_32_1]], [[MOVXM_lng_cg4]]
+    ; CHECK-NEXT: [[VADD_32_:%[0-9]+]]:vec512 = VADD_32 [[VSEL_32_]], [[VSRS_S32_S64_mv_x_srs]]
+    ; CHECK-NEXT: [[COPY15:%[0-9]+]]:ewl = COPY [[VADD_32_]].sub_256_lo
+    ; CHECK-NEXT: [[COPY16:%[0-9]+]]:ewh = COPY [[VADD_32_]].sub_256_hi
+    ; CHECK-NEXT: [[VLDB_4x16_LO:%[0-9]+]]:vec256 = VLDB_4x16_LO [[COPY15]] :: (load unknown-size from @softmax_ilut_ab, align 1), (load unknown-size from @softmax_ilut_cd, align 1)
```

Comment:
I like it! I'm just wondering how safe we can guarantee this is? I think the 4x loads are ultimately touching the 4 global buffers. But I guess we can't really say that using MMOs?

Reply:
Hi @gbossu, this test is an extraction of Softmax. In this case, we have 4 pointers derived from two globals. If we have more globals we can track as well. From the ISA perspective, the only restriction is that pointers 0 and 2 should access even banks and 1 and 3 odd banks. As we are not feeding size and offset information, the only information that AA can use to analyze is the pointer underlying object (always mustAlias for same underlying objects).

Reply:
I'm just wondering about safety, if the load can touch multiple buffers, the MMO should probably reflect that. Because we then need to check each of those locations agains another pointer when doing AA. Maybe that means we need a new PseudoSourceValue to represent a union of locations?

Reply:
Don't we/AA get the address space from the MMO? we could add the combined one then.

Reply:
Hi @gbossu, I think I forgot to give a more comprehensive answer here a, sorry! Considering multiple MMOS, is safe to add as much as needed. For example, the `MachineInstr::mayAlias` will check all MMOS between two instructions:

```
  // Check each pair of memory operands from both instructions, which can't
  // alias only if all pairs won't alias.
  for (auto *MMOa : memoperands())
    for (auto *MMOb : Other.memoperands())
      if (MemOperandsHaveAlias(MFI, AA, UseTBAA, MMOa, MMOb))
        return true;

  return false;
```

In this case, as we are adding all or nothing, it will be safe.